python/lsst/daf/butler/registry/interfaces/_collections.py

```diff
@@ -50,7 +49,10 @@
     from ._dimensions import DimensionRecordStorageManager
 
 
-class CollectionRecord:
+_Key = TypeVar("_Key")
+
+
+class CollectionRecord(Generic[_Key]):
```

Comment:
This is probably a topic for another ticket, but I'm guessing we'll find the typing works ever better (especially if we end up sending these things from server to client via pydantic) if we switch from an ABC pattern to a descriminated union one, e.g.:
```
@dataclasses.dataclass
class RunRecord(Generic[_K]):
    key: _K
    name: str
    type: Literal[CollectionType.RUN]

@dataclasses.dataclass
class ChainedCollectionRecord(Generic[_K]):
    key: _K
    name: str
    type: Literal[CollectionType.CHAINED]
    children: Sequence[str]

CollectionRecord: TypeAlias = Union[RunRecord, ChainedCollectionRecord]   # etc.
```

But :+1: to adding generics here.

Reply:
We'll probably need a `TaggedCollectionRecord` also?

Reply:
Yes, and `CalibrationCollectionRecord`.  Just didn't try to write out the whole thing.

Reply:
Looks reasonable, I think this deserves a dedicated ticket.